File: src/main/java/net/wyvest/hytilities/handlers/chat/modules/triggers/AutoQueue.java
/*
 * Hytilities Reborn - Hypixel focused Quality of Life mod.
 * Copyright (C) 2021  W-OVERFLOW
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package net.wyvest.hytilities.handlers.chat.modules.triggers;

import net.wyvest.hytilities.Hytilities;
import net.wyvest.hytilities.config.HytilitiesConfig;
import net.wyvest.hytilities.handlers.chat.ChatReceiveModule;
import net.wyvest.hytilities.handlers.game.GameType;
import net.wyvest.hytilities.handlers.language.LanguageData;
import gg.essential.api.utils.Multithreading;
import net.minecraft.util.IChatComponent;
import net.minecraftforge.client.event.ClientChatReceivedEvent;
import net.minecraftforge.event.world.WorldEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.InputEvent;
import net.wyvest.hytilities.util.locraw.LocrawInformation;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

public class AutoQueue implements ChatReceiveModule {

    /**
     * We want this to activate early so that it catches the queue message.
     */
    @Override
    public int getPriority() {
        return -11;
    }

    private String command = null;
    private boolean sentCommand;

    @Override
    public void onMessageReceived(@NotNull ClientChatReceivedEvent event) {
        if (!HytilitiesConfig.autoQueue) {
            return;
        }

        final LanguageData language = getLanguage();
        final String message = getStrippedMessage(event.message);
        if (message.startsWith(language.autoQueuePrefix)) {
            for (IChatComponent component : event.message.getSiblings()) {
                final String compMsg = getStrippedMessage(event.message).trim();
                if (compMsg.equals(language.autoQueueClick)) { // TODO: HELP DOES NOT WORK
                    this.command = component.getChatStyle().getChatClickEvent().getValue();
                }
            }
        }
        LocrawInformation locraw = Hytilities.INSTANCE.getLocrawUtil().getLocrawInformation();
        if (locraw != null && locraw.getGameType() == GameType.BED_WARS) {
            if (message.startsWith(language.autoQueuePrefixBedwars)) {
                switch (locraw.getGameMode()) {
                    case "BEDWARS_EIGHT_ONE": // Solo
                        this.command = "/play bedwars_eight_one";
                    case "BEDWARS_EIGHT_TWO": // Doubles
                        this.command = "/play bedwars_eight_two";
                    case "BEDWARS_FOUR_THREE": // Threes
                        this.command = "/play bedwars_four_three";
                    case "BEDWARS_FOUR_FOUR": // Fours
                        this.command = "/play bedwars_four_four"; // 4v4
                    case "BEDWARS_EIGHT_TWO_VOIDLESS": // Voidless Doubles
                        this.command = "/play bedwars_eight_two_voidless";
                    case "BEDWARS_FOUR_FOUR_VOIDLESS": // Voidless Fours
                        this.command = "/play bedwars_four_four_voidless";
                }
            }
        }
    }

    @Override
    public boolean isEnabled() {
        return HytilitiesConfig.autoQueue;
    }

    @SubscribeEvent
    public void onKeyInput(InputEvent.KeyInputEvent event) {
        if (this.command != null) {
            this.switchGame();
        }
    }

    @SubscribeEvent
    public void onMouseEvent(InputEvent.MouseInputEvent event) {
        if (this.command != null) {
            this.switchGame();
        }
    }

    @SubscribeEvent
    public void onWorldLoad(WorldEvent.Load event) {
        // stop the command from being spammed, to prevent chat from filling with "please do not spam commands"
        if (event.world.provider.getDimensionId() == 0 && this.sentCommand) {
            this.sentCommand = false;
        }
    }

    private void switchGame() {
        Multithreading.schedule(() -> {
            if (!this.sentCommand) {
                Hytilities.INSTANCE.getCommandQueue().queue(this.command);
                this.sentCommand = true;
                this.command = null;
            }

        }, HytilitiesConfig.autoQueueDelay, TimeUnit.SECONDS);
    }
}

File: src/main/kotlin/org/polyfrost/hytils/client/handlers/chat/modules/triggers/AutoQueue.kt
package org.polyfrost.hytils.client.handlers.chat.modules.triggers

import dev.deftu.omnicore.api.client.chat.OmniClientChatSender
import dev.deftu.omnicore.api.world.OmniDimension
import dev.deftu.omnicore.api.world.dimensionType
import net.minecraft.client.multiplayer.ClientLevel
import org.polyfrost.hytils.client.HytilsRebornConfig
import org.polyfrost.hytils.client.data.providers.LanguageData
import org.polyfrost.hytils.client.events.ChatReceiveEvent
import org.polyfrost.hytils.client.data.providers.GameIdentifiersData
import org.polyfrost.hytils.client.handlers.chat.ChatReceiveModule
import org.polyfrost.oneconfig.api.event.v1.events.KeyInputEvent
import org.polyfrost.oneconfig.api.event.v1.events.MouseInputEvent
import org.polyfrost.oneconfig.api.event.v1.events.WorldEvent
import org.polyfrost.oneconfig.api.event.v1.invoke.impl.Subscribe
import org.polyfrost.oneconfig.api.hypixel.v1.HypixelUtils
import org.polyfrost.oneconfig.utils.v1.Multithreading
import java.util.concurrent.TimeUnit

object AutoQueue : ChatReceiveModule {
    private var command: String? = null
    private var sentCommand = false
    private var gameEnded = false

    override fun onChatReceived(event: ChatReceiveEvent) {
        if (event.isOverlay || !hasGameEnded(event.unformattedMessage)) return

        val location = HypixelUtils.getLocation()
        if (location.gameType.isPresent && location.mode.isPresent) {
            val databaseName = location.gameType.get().databaseName
            val gameMode = location.mode.get()
            val identifier = GameIdentifiersData.games[databaseName]?.get(gameMode) ?: gameMode
            command = "/play $identifier"
        }
    }

    @Subscribe
    fun onKeyInput(event: KeyInputEvent) {
        if (HytilsRebornConfig.isEnabled && command != null) {
            queue()
        }
    }

    // TODO: mouse input doesn't seem to work
    @Subscribe
    fun onMouseInput(event: MouseInputEvent) {
        if (HytilsRebornConfig.isEnabled && command != null) {
            queue()
        }
    }

    @Subscribe
    fun onWorldLoad(event: WorldEvent.Load) {
        gameEnded = false

        // stop the command from being spammed
        if (event.getWorld<ClientLevel>().dimensionType == OmniDimension.Overworld && sentCommand) {
            sentCommand = false
        }
    }

    private fun queue() {
        Multithreading.schedule({
            if (!sentCommand) {
                OmniClientChatSender.send(command!!) // FIXME: .queue
                sentCommand = true
                command = null
            }
        }, HytilsRebornConfig.autoQueueDelay.toLong(), TimeUnit.SECONDS)
    }

    private fun hasGameEnded(message: String): Boolean {
        if (gameEnded) return false

        val matched = message.matches(LanguageData.GAME_END)
        if (matched) {
            gameEnded = true
        }

        return matched
    }

    override fun isEnabled() = HytilsRebornConfig.autoQueue
    // this should run earlier than other modules but after autogg
    override fun getPriority() = -2
}
